lisa/sut_orchestrator/qemu/platform.py

```diff
@@ -0,0 +1,447 @@
+        on_crash.text = "destroy"
+
+        devices = ET.SubElement(domain, "devices")
+
+        serial = ET.SubElement(devices, "serial")
+        serial.attrib["type"] = "pty"
+
+        serial_target = ET.SubElement(serial, "target")
+        serial_target.attrib["type"] = "isa-serial"
+        serial_target.attrib["port"] = "0"
+
+        serial_target_model = ET.SubElement(serial_target, "model")
+        serial_target_model.attrib["name"] = "isa-serial"
+
+        console = ET.SubElement(devices, "console")
+        console.attrib["type"] = "pty"
+
+        console_target = ET.SubElement(console, "target")
+        console_target.attrib["type"] = "serial"
+        console_target.attrib["port"] = "0"
+
+        video = ET.SubElement(devices, "video")
+
+        video_model = ET.SubElement(video, "model")
+        video_model.attrib["type"] = "qxl"
+
+        graphics = ET.SubElement(devices, "graphics")
+        graphics.attrib["type"] = "spice"
+
+        network_interface = ET.SubElement(devices, "interface")
+        network_interface.attrib["type"] = "network"
+
+        network_interface_source = ET.SubElement(network_interface, "source")
+        network_interface_source.attrib["network"] = "default"
+
+        cloud_init_disk = ET.SubElement(devices, "disk")
+        cloud_init_disk.attrib["type"] = "file"
+        cloud_init_disk.attrib["device"] = "cdrom"
+
+        cloud_init_disk_driver = ET.SubElement(cloud_init_disk, "driver")
+        cloud_init_disk_driver.attrib["name"] = "qemu"
+        cloud_init_disk_driver.attrib["type"] = "raw"
+
+        cloud_init_disk_target = ET.SubElement(cloud_init_disk, "target")
+        cloud_init_disk_target.attrib["dev"] = "sda"
+        cloud_init_disk_target.attrib["bus"] = "sata"
+
+        cloud_init_disk_target = ET.SubElement(cloud_init_disk, "source")
+        cloud_init_disk_target.attrib["file"] = node_context.cloud_init_file_path
+
+        os_disk = ET.SubElement(devices, "disk")
+        os_disk.attrib["type"] = "file"
+        os_disk.attrib["device"] = "disk"
+
+        os_disk_driver = ET.SubElement(os_disk, "driver")
+        os_disk_driver.attrib["name"] = "qemu"
+        os_disk_driver.attrib["type"] = "qcow2"
+
+        os_disk_target = ET.SubElement(os_disk, "target")
+        os_disk_target.attrib["dev"] = "sdb"
+        os_disk_target.attrib["bus"] = "sata"
+
+        os_disk_target = ET.SubElement(os_disk, "source")
+        os_disk_target.attrib["file"] = node_context.os_disk_file_path
+
+        xml = ET.tostring(domain, "unicode")
+        return xml
+
+    # Wait for the VM to boot and then get the IP address.
+    def _get_node_ip_address(
+        self,
+        environment: Environment,
+        log: Logger,
+        qemu_conn: libvirt.virConnect,
+        node: Node,
+        timeout: float,
+    ) -> str:
+        node_context = get_node_context(node)
+
+        while True:
+            addr = self._try_get_node_ip_address(environment, log, qemu_conn, node)
```

Comment:
There are retry and timeout module, which can be used here. Search `@retry` or `@func_set_timeout` to learn from code. Note, if the timeout is used, make sure to capture the timeout exception separately like other usages.

Reply:
In my current code, a single timeout time is shared between all the nodes. This stops the total timeout length from increasing with the number of nodes.